File: lib/socket_path.cpp
#include "inspector/socket_path.h"

#include <cstdlib>
#include <filesystem>

#ifdef _WIN32
#include <windows.h>
#elif defined(__APPLE__) || defined(__linux__)
#include <unistd.h>
#endif

namespace inspector {

std::filesystem::path getSocketDirectory() {
#ifdef __linux__
  // On Linux, prefer XDG_RUNTIME_DIR (usually /run/user/UID)
  const char *xdgRuntimeDir = std::getenv("XDG_RUNTIME_DIR");
  if (xdgRuntimeDir) {
    return std::filesystem::path(xdgRuntimeDir) / "inspector";
  }
  // Fallback to ~/.cache/inspector
  const char *home = std::getenv("HOME");
  if (home) {
    return std::filesystem::path(home) / ".cache" / "inspector";
  }
  // Should never happen, but use current directory as last resort
  return std::filesystem::path(".") / ".inspector";

#elif defined(__APPLE__)
  // On macOS, use ~/Library/Application Support for persistent sockets
  const char *home = std::getenv("HOME");
  if (home) {
    return std::filesystem::path(home) / "Library" / "Application Support" /
           "inspector";
  }
  // Fallback to ~/.cache
  return std::filesystem::path(".") / ".inspector";

#elif defined(_WIN32)
  // On Windows, use %LOCALAPPDATA% for persistent app data
  const char *localAppData = std::getenv("LOCALAPPDATA");
  if (localAppData) {
    return std::filesystem::path(localAppData) / "inspector";
  }
  // Fallback to %APPDATA%
  const char *appData = std::getenv("APPDATA");
  if (appData) {
    return std::filesystem::path(appData) / "inspector";
  }
  // Fallback to user profile
  const char *userProfile = std::getenv("USERPROFILE");
  if (userProfile) {
    return std::filesystem::path(userProfile) / ".inspector";
  }
  return std::filesystem::path(".") / ".inspector";

#else
  // Generic Unix fallback: use ~/.cache
  const char *home = std::getenv("HOME");
  if (home) {
    return std::filesystem::path(home) / ".cache" / "inspector";
  }
  return std::filesystem::path(".") / ".inspector";
#endif
}

} // namespace inspector

File: python/inspector/socket_path.py
#!/usr/bin/env python

import os
import platform
from pathlib import Path


def get_socket_directory() -> Path:
    """Get platform-specific socket directory."""
    system = platform.system()

    if system == "Linux":
        # On Linux, prefer XDG_RUNTIME_DIR (usually /run/user/UID)
        xdg_runtime_dir = os.environ.get("XDG_RUNTIME_DIR")
        if xdg_runtime_dir:
            return Path(xdg_runtime_dir) / "inspector"
        # Fallback to ~/.cache/inspector
        home = os.environ.get("HOME")
        if home:
            return Path(home) / ".cache" / "inspector"
        # Should never happen, but use current directory as last resort
        return Path(".") / ".inspector"

    elif system == "Darwin":  # macOS
        # On macOS, use ~/Library/Application Support for persistent sockets
        home = os.environ.get("HOME")
        if home:
            return Path(home) / "Library" / "Application Support" / "inspector"
        # Fallback to current directory
        return Path(".") / ".inspector"

    elif system == "Windows":
        # On Windows, use %LOCALAPPDATA% for persistent app data
        local_app_data = os.environ.get("LOCALAPPDATA")
        if local_app_data:
            return Path(local_app_data) / "inspector"
        # Fallback to %APPDATA%
        app_data = os.environ.get("APPDATA")
        if app_data:
            return Path(app_data) / "inspector"
        # Fallback to user profile
        user_profile = os.environ.get("USERPROFILE")
        if user_profile:
            return Path(user_profile) / ".inspector"
        return Path(".") / ".inspector"

    else:
        # Generic Unix fallback: use ~/.cache
        home = os.environ.get("HOME")
        if home:
            return Path(home) / ".cache" / "inspector"
        return Path(".") / ".inspector"
